nav2_pure_pursuit_controller/src/pure_pursuit_controller.cpp

```diff
@@ -85,19 +108,34 @@ geometry_msgs::msg::TwistStamped PurePursuitController::computeVelocityCommands(
   const geometry_msgs::msg::PoseStamped & pose,
   const geometry_msgs::msg::Twist &)
 {
-  // Find the first pose which is at a distance greater than the specified lookahed distance
-  auto goal_pose = std::find_if(
-    global_plan_.poses.begin(), global_plan_.poses.end(),
-    [&](const auto & global_plan_pose) {
-      return hypot(
-        global_plan_pose.pose.position.x,
-        global_plan_pose.pose.position.y) >= lookahead_dist_;
-    })->pose;
+  auto transformed_plan = transformGlobalPlan(pose);
+
+  // First find the closest pose on the path to the robot
+  auto closest_pose_it =
```

Comment:
Why not use the look ahead distance variable? You do not simply want the closest one.

Reply:
I don't really see the point of separating this into 2 lambdas - can you explain it?

Reply:
I hope this figure is making it clear:
![nav2 bug explained (1)](https://user-images.githubusercontent.com/91969/98924511-345bd680-24d5-11eb-946d-04142a7b17cf.png)

Reply:
Ah I see, I've fixed this before in other controllers by removing points from the path as it passes it. So if a point has been passed by the robot, its culled from the path stored in the global_path_ variable so that only current/future points are available. DWA did this as well. Then the look ahead point can only be from the forward motion

Reply:
But I think this is OK then for the tutorial code, but it would be more efficient with those changes if you were actually using this on your robot platform.

Reply:
As an aside, are you using this on your platform? If so, do you think this should be moved out of tutorials and into the main nav2 repo? Any changes you'd want to see made for that?'

That figure is nice as well. If you remove the "previously this point.." arrow, you could upload that to this PR and package README to explain how it works for users. Very useful.

Reply:
I could implement path pruning, but then I would copy the implementation from DWB. Could we design/move this code to a utility library so this package can also use that? I'm also fine with just copying the DWB pruning code in here.

I'm not using this on a platform, but I was very confused why the code from the tutorial didn't work. I think it's better to keep this controller as simple as possible as a tutorial.

I'll update this PR with the figure

Reply:
Lets just copy it, especially if its just staying as tutorial code.

Reply:
I've copied and adapted the path pruning code from dwb to this package. See the updated gif in the PR. I've changed it a little bit to prune towards the closest point to the robot. This works better with determining the lookahead point.

I've also added a small README.

Let me know what you think.